### PR TITLE
Issue-6755: Additional Firefox versions for Build Addons

### PR DESCRIPTION
**Pull request by jrbenny35**
This is an initial idea of what I had in mind to allow for the downloading of Developer and Nightly versions of Firefox.

Please feel free to comment. I haven't done much ruby coding in my short career.

### CODE REVIEW

File: lib/travis/build/addons/firefox.rb [hunk 1]
```diff
@@ -68,13 +69,32 @@ def export_source_url
               'firefox-beta-latest'
             when 'latest-esr'
               'firefox-esr-latest'
+            when 'latest-dev'
+              'firefox-aurora-latest'
+            when 'latest-nightly'
+              if RbConfig::CONFIG['host_os'] == /linux/
```
Comment (BanzaiMan) [inline lib/travis/build/addons/firefox.rb]: This does not work. The bash script is compiled on a separate machine, which means that the platform detection must be done via the compiled bash script, not by `travis-build`.

Comment (jrbenny35) [inline lib/travis/build/addons/firefox.rb]: Makes sense. Thanks!

**Comment by jrbenny35** [comment]
Could I ask for another review?

File: lib/travis/build/addons/firefox.rb [hunk 2]
```diff
@@ -68,13 +68,27 @@ def export_source_url
               'firefox-beta-latest'
             when 'latest-esr'
               'firefox-esr-latest'
+            when 'latest-dev'
+              'firefox-aurora-latest'
```
Comment (BanzaiMan) [inline lib/travis/build/addons/firefox.rb]: What is `aurora`? Is there a chance that this is going to change? If so, how often does it change, and where should we look for the replacement?

File: lib/travis/build/addons/firefox.rb [hunk 3]
```diff
@@ -68,13 +68,27 @@ def export_source_url
...
-            host = 'download.mozilla.org'
+            if product == 'nightly'
+              host = 'archive.mozilla.org/pub/firefox/nightly/latest-mozilla-central/'
+            else
+              host = 'download.mozilla.org'
+            end
 
-            sh.if "$(uname) = 'Linux'" do
+            sh.if "[$(uname) = 'Linux'] && [#{product} = 'nightly']" do
+              sh.export 'FIREFOX_SOURCE_URL', "'https://#{host}//.+?(?=linux-x86_64)/.linux-x86_64.tar.bz2'"
```
Comment (BanzaiMan) [inline lib/travis/build/addons/firefox.rb]: Does this work?

Comment (jrbenny35) [inline lib/travis/build/addons/firefox.rb]: You're talking about the regular expression?

**Comment by jrbenny35** [comment]
This is the type of link I am trying to get using regEx from the mozilla servers. http://archive.mozilla.org/pub/firefox/nightly/latest-mozilla-central/firefox-52.0a1.en-US.linux-x86_64.tar.bz2

I am trying to think of a way to get this url without someone having to update it each time a new firefox version comes out. If not, I could be the one to update it. It would only be nightly that needs updating.

**Comment by jrbenny35** [comment]
Looking at this again, the only way I think it would work without constant updates would be through the use of a web scraper. I don't know the policies travis holds with regards to adding gems to their ruby build scripts.

**Comment by jrbenny35** [comment]
Did some idea tossing with my colleagues and we came up with a solution using download.mozilla.org. It has a bouncer app for the product search and it works well! This should work 10/10 times unless mozilla changes something.

**Comment by BanzaiMan** [comment]
Hmm. The PR now seems completely invalid. For example, the last commit https://github.com/travis-ci/travis-build/pull/864/commits/4961b6ff320d61f9a9e933387c7e5b3d9967d02b shows me 436 additions and 200 deletions, with no real change.

issue-6755__additional_firefox_versions_for_build_addons_by_jrbenny35_ _pull_request__864_ _travis-ci_travis-build

I don't know what's going on, but I certainly cannot review this.

**Comment by jrbenny35** [comment]
Looks like the merge commits were the last pushed to the branch. Can you just review the commit 	3f89290? Looks like their should've been merge conflicts but nothing came up when I pushed. If not I can start a new PR and branch.

**Comment by BanzaiMan** [comment]
Force push with what you think is the right commit, please.

**Comment by jrbenny35** [comment]
I think I got it, sorry about that.

File: lib/travis/build/addons/firefox.rb (hunk 2 above)
Comment (BanzaiMan) [inline lib/travis/build/addons/firefox.rb]: @jrbenny35 This needs to be addressed before merging this.

Comment (jrbenny35) [inline lib/travis/build/addons/firefox.rb]: Whoops, I thought I replied to this. Aurora is the term for the branch that holds firefox developer edition. Developer edition is usually any features that have been on nightly for ~6 weeks.

Comment (davehunt) [inline lib/travis/build/addons/firefox.rb]: See https://wiki.mozilla.org/Firefox/Channels#Developer_Edition_.28aka_Aurora.29

Comment (jrbenny35) [inline lib/travis/build/addons/firefox.rb]: Thanks Dave! @BanzaiMan that link explains it better.

**Comment by jrbenny35** [comment]
Thanks!

**Comment by jrbenny35** [comment]
@BanzaiMan How soon until I could use this in my project?

**Comment by BanzaiMan** [comment]
This has been deployed (along with my touchups shown above).

**Comment by BanzaiMan** [comment]
Also, we welcome docs PR!

**Comment by jrbenny35** [comment]
I'll get on that!

**Pull request merged**